model-engine/model_engine_server/common/concurrency_limiter.py

```diff
@@ -0,0 +1,35 @@
+from multiprocessing import BoundedSemaphore
+from multiprocessing.synchronize import BoundedSemaphore as BoundedSemaphoreType
+from typing import Optional
+
+from fastapi import HTTPException
+from model_engine_server.core.loggers import logger_name, make_logger
+
+logger = make_logger(logger_name())
+
+
+class MultiprocessingConcurrencyLimiter:
+    def __init__(self, concurrency: Optional[int], fail_on_concurrency_limit: bool):
+        if concurrency is not None:
+            if concurrency < 1:
+                raise ValueError("Concurrency should be at least 1")
+            self.semaphore: Optional[BoundedSemaphoreType] = BoundedSemaphore(value=concurrency)
+            self.blocking = (
+                not fail_on_concurrency_limit
+            )  # we want to block if we want to queue up requests
+        else:
+            self.semaphore = None
+            self.blocking = False  # Unused
+
+    def __enter__(self):
+        logger.debug("Entering concurrency limiter semaphore")
+        if self.semaphore and not self.semaphore.acquire(block=self.blocking):
+            logger.warning("Too many requests, returning 429")
```

Comment:
nit: can we log the # of requests instead of this log? i.e. # of requests that exceed the concurrency limit

Reply:
any requests over the concurrency limit will immediately be returned as 429, so we'd always be logging whatever the value of `MAX_CONCURRENCY` is (which still seems good to log though)